### PR TITLE
add Nano HPO notebooks

### PR DESCRIPTION
**Pull request by guoyaol**
three different methods of using tensorflow

### CODE REVIEW

File: python/nano/notebooks/hpo/custom.ipynb [hunk 1]
```diff
@@ -0,0 +1,1609 @@
...
+      "  warnings.warn(\"NANO_TF_INTER_OP not found the in os.environ, \"\n",
+      "../../src/bigdl/nano/tf/__init__.py:30: UserWarning: OMP_NUM_THREADS not found the in os.environ, please run `source bigdl-nano-init`\n",
+      "  warnings.warn(\"OMP_NUM_THREADS not found the in os.environ, \"\n"
+     ]
+    }
+   ],
+   "source": [
+    "#to do: remove\n",
+    "import sys\n",
+    "sys.path.append(\"../../src/\")\n",
```
Comment (shane-huang) [inline python/nano/notebooks/hpo/custom.ipynb:62]: remove sys.path.append(...)

File: python/nano/notebooks/hpo/custom.ipynb [hunk 2]
```diff
@@ -0,0 +1,1609 @@
...
+    "\n",
+    "pip install --pre --upgrade bigdl"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Step1: Init Nano AutoML\n",
+    "Let's start out by importing Nano AutoML and enable its tensorflow API globally by automl.hpo_config.enable_hpo_tf().\n",
```
Comment (shane-huang) [inline python/nano/notebooks/hpo/custom.ipynb:38]: break this large cell into two steps: one for "Init Nano AutoML" and one for "Prepare Data"

**Comment by shane-huang** [comment]
Try all the notebooks on google colab and see if it works and how long it takes to run.

File: python/nano/notebooks/hpo/custom.ipynb [hunk 3]
```diff
@@ -0,0 +1,1609 @@
...
+    "    optimizer=RMSprop(learning_rate=1e-4),\n",
+    "    metrics=[\"accuracy\"],\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Step3: Apply searching\n",
```
Comment (shane-huang) [inline python/nano/notebooks/hpo/custom.ipynb:154]: Step 3: Run Hyperparamter Search

File: python/nano/notebooks/hpo/custom.ipynb [hunk 4]
```diff
@@ -0,0 +1,1609 @@
...
+      "CPU times: user 18min 5s, sys: 13min 26s, total: 31min 31s\n",
+      "Wall time: 4min 23s\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "model.search(\n",
+    "    n_trials=20,\n",
+    "    target_metric='accuracy',\n",
```
Comment (shane-huang) [inline python/nano/notebooks/hpo/custom.ipynb:1448]: use val_accuracy for target_metric.

File: python/nano/notebooks/hpo/custom.ipynb [hunk 5]
```diff
@@ -0,0 +1,1609 @@
...
+      "19/19 [==============================] - 1s 34ms/step - loss: 0.2048 - accuracy: 0.9454 - val_loss: 0.3218 - val_accuracy: 0.9017\n",
+      "Epoch 22/24\n",
+      "19/19 [==============================] - 1s 35ms/step - loss: 0.1991 - accuracy: 0.9433 - val_loss: 0.3344 - val_accuracy: 0.9000\n",
+      "Epoch 23/24\n",
+      "19/19 [==============================] - 1s 32ms/step - loss: 0.1935 - accuracy: 0.9479 - val_loss: 0.3312 - val_accuracy: 0.9033\n",
+      "Epoch 24/24\n",
+      "19/19 [==============================] - 1s 32ms/step - loss: 0.1875 - accuracy: 0.9504 - val_loss: 0.3425 - val_accuracy: 0.9000\n",
+      "32/32 - 1s - loss: 0.3790 - accuracy: 0.8760 - 832ms/epoch - 26ms/step\n",
+      "Test loss: 0.37904876470565796\n",
+      "Test accuracy: 0.8759999871253967\n"
```
Comment (shane-huang) [inline python/nano/notebooks/hpo/custom.ipynb:1528]: This test accuracy is too low for MNIST. We need some tuning on this one. Use val_accuracy for target metric and try again.

File: python/nano/notebooks/hpo/sequential.ipynb [hunk 6]
```diff
@@ -0,0 +1,1593 @@
...
+      "19/19 [==============================] - 0s 18ms/step - loss: 0.2048 - accuracy: 0.9454 - val_loss: 0.3218 - val_accuracy: 0.9017\n",
+      "Epoch 22/24\n",
+      "19/19 [==============================] - 1s 28ms/step - loss: 0.1991 - accuracy: 0.9433 - val_loss: 0.3344 - val_accuracy: 0.9000\n",
+      "Epoch 23/24\n",
+      "19/19 [==============================] - 1s 28ms/step - loss: 0.1935 - accuracy: 0.9479 - val_loss: 0.3312 - val_accuracy: 0.9033\n",
+      "Epoch 24/24\n",
+      "19/19 [==============================] - 1s 27ms/step - loss: 0.1875 - accuracy: 0.9504 - val_loss: 0.3425 - val_accuracy: 0.9000\n",
+      "32/32 - 1s - loss: 0.3790 - accuracy: 0.8760 - 722ms/epoch - 23ms/step\n",
+      "Test loss: 0.37904876470565796\n",
+      "Test accuracy: 0.8759999871253967\n"
```
Comment (shane-huang) [inline python/nano/notebooks/hpo/sequential.ipynb:1512]: This accuracy is too low for MNIST. We may change the network architecture or you can use some MNIST architecture from other examples.

**Comment by shane-huang** [comment]
1. split this PR into two PRs
     - one demo shows two ways of defining models (functional and sequential) in one notebook. The network structure can be the same or not. 
     - one demo shows defining model by subclassing an existing model. 
2. submit a PR for pytorch HPO. Modify from this pytorch lightning example https://pytorch-lightning.readthedocs.io/en/latest/notebooks/lightning_examples/cifar10-baseline.html. For how to use pytorch HPO, learn from the pytorch unit test (https://github.com/intel-analytics/BigDL/blob/main/python/nano/test/automl/pytorch/test_trainer.py)

**Comment by shane-huang** [comment]
rerun the failed check and until you get it pass. Or the PR can't be merged.

**Comment by shane-huang** [comment]
1. >In this guide we will demonstrate how to use Nano AutoML to auto tune an machine learning model and handle the whole model development process easily.

Delete this line. 

1. > Nano provides hpo as an solution for fast and high-performance hyperparameter tuning. Users can import hano hpo following the standard tensorflow API without changing their own code. By setting the search space and running model.search(), the model will contain the best hyperparameters.
   >
    > Nano hpo supports sequential, fuctional and custom keras API.

In this notebook we demonstrates how to use Nano HPO to tune the hyperparameters in tensorflow training. The models are built using tensorflow keras sequential API and functional API.

**Comment by shane-huang** [comment]
The content in `Step0: Prepare Environment` is not correct, please follow the user guide. basically we first install bigdl-nano (https://bigdl.readthedocs.io/en/latest/doc/Nano/Overview/nano.html#install), then install some dependences as instructed at https://bigdl.readthedocs.io/en/latest/doc/Nano/QuickStart/hpo.html#install

**Comment by shane-huang** [comment]
Make the notebook runnable in google colab. Finally we would like something similar to chronos tutorial page ( https://bigdl.readthedocs.io/en/latest/doc/Chronos/QuickStart/index.html)

**Comment by shane-huang** [comment]
1. >Let's start out by importing Nano AutoML and enable its tensorflow API globally by automl.hpo_config.enable_hpo_tf().

We need to enable it before we use Nano HPO for tensorflow training.

2. > After importing, we will load MNIST as our data.

We use MNIST dataset for demonstration.

**Comment by shane-huang** [comment]
1. >Apply hyperparameter searching by calling model.search(), and best hyperparameters will be saved in the model.
   >You can specify the number of trial, each trial will use a set of hyperparameters automatiacally suggusted by nano.hpo.
   >You can set your target metric and whehter you want to maximize you minimize it. if you set study_name and storage path, you can save the search history in .db file and resume it when you need it.

Run hyperparameter search by calling `model.search`. Set `n_trials` to the number of trails you want to run, and set the `target_metric` and `direction` so that HPO optimizes the `target_metric` in the specified `direction`. Each trial will use a different set of hyperparameters in the search space range. Use `n_parallels` to set the nubmer of parallel processes to run trials. For more details, refer to [..link to user doc]

**Comment by shane-huang** [comment]
> if you set resume = True, you can specify study name and storage path. You will be able to load the search history in .db file and resume it for the next trial.

You can resume the previous search when a search completes by setting `resume=True`. Refer to [user doc] for more details.

**Comment by shane-huang** [comment]
>The best hyperparameters have been saved in the model. You can call model.fit() to train a model with the best hyperparameters.

After search, `model.fit` will autotmatically use the best hyperparmeters found in search for the fitting the model.

**Comment by shane-huang** [comment]
Add a section 8 for HPO Result Analysis and Visualization

**Comment by shane-huang** [comment]
submit a PR to bigdl to allow users to use bigdl.nano.automl.hpo.visualization.* instead of optuna.visualization for visualization. 
revise the notebook to use bigdl instead of optuna.

**Comment by shane-huang** [comment]
run in colab.

**Comment by shane-huang** [comment]
-  Seperate bigdl-nano install and HPO dependences into two blocks. 

- Change to below commands to install bigdl-nano.  Change the text to "You can install the latest pre-release version with nano support using below commands". Add a sentence "We recommend to run below commands, especially `source bigdl-nano-init` before jupyter kernel is started, or some of the optimizations may not take effect".
```bash
pip install --pre bigdl-nano[tensorflow]
pip install setuptools==58.0.4
pip install protobuf==3.20.1
source bigdl-nano-init
```
- Remove block 5 and "After that, we need to import nano and other dependencies. " 
- Add below import at the top of block 6
```python
from tensorflow import keras
``` 
- Add a new block as below for imports before block 7
```
from bigdl.nano.automl.tf.keras import Sequential
from bigdl.nano.tf.keras.layers import Dense, Flatten, Conv2D
from bigdl.nano.tf.keras import Input
from bigdl.nano.automl.tf.keras import Model
import bigdl.nano.automl.hpo.space as space
```
- >Whenever you want to search a hyperparameter, you can simply replace it with a space and set your own searching space.

change to: Change the imports from `tensorflow.keras` to `bigdl.nano` as below, and you will be able to specify search spaces as you define the model. For how to specify search space, refer to [user guide](link).  

- > (Optional)By running the next block, you can build this model using tensorflow keras sequential API.

change to: Below two cells show how to define the model with search spaces using either sequential or functional API respectively.  You can choose one of them to run. 

- >(Optional)Or you can build this model by functional API.

delete it. 

- block 7 and the next should be the same model but wrote in sequential or functional API. Now you used two different models, the user might get confused. 
- add the below imports before block 8
```python
from tensorflow.keras.optimizers import RMSprop
```
- add below imports before block 9
```python
from bigdl.nano.automl.hpo.backend import PrunerType
```

**Comment by shane-huang** [comment]
* cell 15, you don't need to use print, just use below and notebook will show a better table format. 
```python
study.trials_dataframe(attrs=("number", "value", "params", "state"))
```

**Comment by shane-huang** [comment]
In custom.py
* just put `!source bigdl-nano-init` in cell[1]. 
* Do we need `exit()` in cell[2]
* Change below line
> Change the imports from tensorflow.keras to bigdl.nano as below, and you will be able to specify search spaces as you define the model. For how to specify search space, refer to [user doc](https://bigdl.readthedocs.io/en/latest/doc/Nano/QuickStart/hpo.html).

to 

> Decorate the model class with `hpo.tfmodel`, and you will be able to specify search spaces in init arguments when creating the model, as shown below. For more details, refer to [user doc](https://bigdl.readthedocs.io/en/latest/doc/Nano/QuickStart/hpo.html)
* Remove this line `The cell below shows how to define the model with search spaces using custom class.`
* `Step3: Set search space` -> `Step3: Build model and specify search spaces`

**Comment by shane-huang** [comment]
In seq_func.ipynb
* just put `!source bigdl-nano-init` in cell[1]. 
* Do we need `exit()` in cell[2]
* Change `Step3: Set search space` to `Step3: Build model and specify search spaces`
* Change below line
>We now compile our model with loss function, optimizer and metrics.` 

to 

>We now compile our model with loss function, optimizer and metrics. If you want to tune learning rate and batch size, refer to [user guide](https://bigdl.readthedocs.io/en/latest/doc/Nano/QuickStart/hpo.html#search-the-learning-rate)
* Remove this sentence `Use n_parallels to set the nubmer of parallel processes to run trials.` in Step 5 text.

**Comment by shane-huang** [comment]
In seq_func.ipynb
* just put `!source bigdl-nano-init` in cell[1]. 
* Do we need `exit()` in cell[2]
* Change `Step3: Set search space` to `Step3: Build model and specify search spaces`
* Change below line
>We now compile our model with loss function, optimizer and metrics.` 

to 

>We now compile our model with loss function, optimizer and metrics. If you want to tune learning rate and batch size, refer to [user guide](https://bigdl.readthedocs.io/en/latest/doc/Nano/QuickStart/hpo.html#search-the-learning-rate)
* Remove this sentence `Use n_parallels to set the nubmer of parallel processes to run trials.` in Step 5 text.

**Comment by yangw1234** [comment]
I think we need tests for these notebooks.

see examples here https://github.com/intel-analytics/BigDL/blob/main/.github/workflows/nano_notebooks_tests.yml#L40-L53
